### PR TITLE
Chisel prelude files

### PR DESCRIPTION
**Pull request by sug0**
## Motivation

Currently (AFAIK) there is no way to load arbitrary Solidity files in Chisel, when we boot up the repl. Prelude files are an excellent way to define application specific libraries of code to use in manual tests.

## Solution

A new CLI argument was added to Chisel to parse and evaluate prelude Solidity files from a given directory, before reaching the top-level of the REPL.

### CODE REVIEW

File: crates/chisel/bin/main.rs [hunk 1]
```diff
@@ -46,6 +48,10 @@ pub struct ChiselParser {
     #[command(subcommand)]
     pub sub: Option<ChiselParserSub>,
 
+    /// Path to a Solidity file to import as a prelude
+    #[clap(long, help_heading = "REPL options")]
+    pub prelude: Option<PathBuf>,
```
Comment (mattsse) [inline crates/chisel/bin/main.rs:53]: should this be a Vec instead? so multiple files can be loaded?

maybe even a dir?

**Review by mattsse — CHANGES_REQUESTED** [review]
supportive, would like to see support for multiple args

File: crates/chisel/bin/main.rs (hunk 1 above)
Comment (sug0) [reply crates/chisel/bin/main.rs]: added a dir and some spiffy log msgs

File: crates/chisel/bin/main.rs [hunk 2]
```diff
@@ -228,3 +228,48 @@ impl Provider for ChiselParser {
...
+}
+
+/// Evaluate multiple Solidity source files contained within a
+/// Chisel prelude directory.
+async fn evaluate_prelude(
+    dispatcher: &mut ChiselDispatcher,
+    maybe_prelude: Option<PathBuf>,
+) -> eyre::Result<()> {
+    if let Some(prelude_dir) = maybe_prelude {
+        let prelude_sources = fs::files_with_ext(prelude_dir, "sol");
```
Comment (Evalir) [inline crates/chisel/bin/main.rs:256]: Hmm, might make things just a tad more complicated, but wdyt about supporting both:
- Loading a single file
- Pointing to a directory of files

If you find it blows up the complexity of the command too much, feel free to ignore—just think a lot of ppl would appreciate just putting a file somewhere and loading it

File: crates/chisel/bin/main.rs [hunk 3]
```diff
@@ -228,3 +228,48 @@ impl Provider for ChiselParser {
...
+        let print_success_msg = !prelude_sources.is_empty();
+
+        for source_file in prelude_sources {
+            println!(
+                "{} {}",
+                Paint::yellow("⏳ Loading prelude source file:"),
```
Comment (Evalir) [inline crates/chisel/bin/main.rs:262]: I'd say let's avoid emojis on the output—might not display accurately on all platforms!

File: crates/chisel/bin/main.rs [hunk 4]
```diff
@@ -228,3 +228,48 @@ impl Provider for ChiselParser {
...
+                source_file.display(),
+            );
+
+            let prelude = fs::read_to_string(source_file)?;
+            dispatch_repl_line(dispatcher, &prelude).await;
+        }
+
+        if print_success_msg {
+            println!("{}\n", Paint::green("✅ All prelude source files loaded successfully"));
```
Comment (Evalir) [inline crates/chisel/bin/main.rs:271]: ditto

**Review by Evalir — CHANGES_REQUESTED** [review]
love this—just a few comments

File: crates/chisel/bin/main.rs (hunk 2 above)
Comment (sug0) [reply crates/chisel/bin/main.rs:256]: I originally just had a single source file supported. @mattsse asked to support multiple files, which I liked... I thought of supporting both files and directories, but it's a bit trickier. I'll see if I find the time to work on that tomorrow.

for my particular use case, a single source file is actually enough. it's possible append multiple `import` statements to a single source file, but then you lose the dynamic nature of not needing to specify which files to import code from

Comment (Evalir) [reply crates/chisel/bin/main.rs:256]: no worries @sug0 — did not mean to conflict with mattsse at all! just would love to support both if possible. happy to help getting this over the line if it becomes too much work :)

**Comment by Evalir** [comment]
friendly ping @sug0 — would love to take this over the line

**Comment by sug0** [comment]
@Evalir sorry, I've been a bit busy. feel free to take over this PR!

**Review by Evalir — APPROVED** [review]
took over this, can load both paths to folders and simple files now.

**Comment by sug0** [comment]
thank you :D

**Review by mattsse — APPROVED** [review]
lgtm, applied 1 style nit

**Comment by mds1** [comment]
Am I using this feature wrong, or is it broken? No matter what I try I see a ton of errors logged, then it says the things were successful. For example, trying to load in forge-std with `chisel --prelude "lib/forge-std/src/Test.sol"` I get this: https://app.warp.dev/block/yeCL5RzCbdShEw28Mn3ixq

I've also tried `"lib/forge-std"`, `"lib/forge-std/src"`, and `"."` as the paths and all give the same thing. In all cases, I cannot access the `makeAddr` cheat from forge-std

**Comment by Evalir** [comment]
Interesting—this should be working albeit it probably will struggle with reading files that import other files. Will test with forge-std and fix @mds1

**Comment by mds1** [comment]
Thanks, yea in general I think it would be great to be able to have the chisel repl inherit from forge-std's `Script` or `Test` contracts. I figured the prelude would be one way to do that, but since that may be a common use case perhaps a dedicated flag for it would be better

**Pull request merged**